File: buckpal/src/main/java/study/architecture/buckpal/account/application/service/GetAccountBalanceService.java
package study.architecture.buckpal.account.application.service;

import java.time.LocalDateTime;
import lombok.RequiredArgsConstructor;
import study.architecture.buckpal.account.application.port.in.GetAccountBalanceQuery;
import study.architecture.buckpal.account.application.port.out.LoadAccountPort;
import study.architecture.buckpal.account.domain.Account;
import study.architecture.buckpal.account.domain.Account.AccountId;
import study.architecture.buckpal.account.domain.Money;

@RequiredArgsConstructor
public class GetAccountBalanceService implements GetAccountBalanceQuery {

  private final LoadAccountPort loadAccountPort;

  @Override
  public Money getAccountBalance(AccountId accountId) {
    Account account = loadAccountPort.loadAccount(accountId, LocalDateTime.now());
    return account.calcualteBalance();
  }
}

File: buckpal-kotlin/src/main/kotlin/study/cleanarchitecture/buckpalkotlin/account/application/service/GetAccountBalanceService.kt
package study.cleanarchitecture.buckpalkotlin.account.application.service

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import study.cleanarchitecture.buckpalkotlin.account.application.port.`in`.GetAccountBalanceQuery
import study.cleanarchitecture.buckpalkotlin.account.application.port.out.LoadAccountPort
import study.cleanarchitecture.buckpalkotlin.account.domain.AccountId
import study.cleanarchitecture.buckpalkotlin.account.domain.Money
import java.time.LocalDateTime

@Service
class GetAccountBalanceService(
    private val loadAccountPort: LoadAccountPort,
) : GetAccountBalanceQuery {

    @Transactional(readOnly = true)
    override fun getAccountBalance(accountId: AccountId): Money {
        return loadAccountPort.loadAccount(accountId, LocalDateTime.now())
            ?.calculateBalance() ?: throw IllegalArgumentException("Can't find account")
    }
}